Guardian change

aw-west opened the pull request:
Guardian link has changed.
I found on `line 195` changing `div.immersive-main-media.immersive-main-media__gallery` to `div.u-responsive-ratio` fixed the issue (today at least)

raffam commented:
I'm sorry but downloaing The Guardian picture works for me as of today. I just tried. 

I also tried to go to [this](https://www.theguardian.com/artanddesign/gallery/2019/jun/29/the-20-photographs-of-the-week) page (which is the page the script obtains).

`div.immersive-main-media.immersive-main-media__gallery` matches the top picture, while your code matches every picture in the page, right?

aw-west commented:
I made a table of my tests.  `patch-2` works for me, I am manually cleaning the `pictures/potd` directory then running `python potd.py --site guardian`.

| Version | master | patch-2 | `div_img` |
| --- | --- | --- | --- |
| 2019-06-29 | Not Tested | [approximate][1b] | [approximate][1c] |
| 2019-06-30<br>Morning | AssertionError line 197<br>`assert(len(img_el) == 1)`<br>`img_el = []` | [`img_url`][2b] | [`div_img`][2c] |
| 2019-06-30<br>Afternoon | [`img_url`][3a] | [`img_url`][3b] | [`div_img`][3c] |

It seems to be a page selection problem.  _Morning_ doesn't have a cover photo, but _Afternoon_ does.  Maybe `try` and `except`, or accept that some times you won't get the headline article photo.
```
try:
    img_el = soup_potd.select("div.immersive-main-media.immersive-main-media__gallery")
except:
    img_el = soup_potd.select("div.u-responsive-ratio")
```

[1b]:https://i.guim.co.uk/img/media/1e469a09541aa3af35052d87b5a9f3b602424a59/0_45_5472_3555/master/5472.jpg?width=1010&quality=85&auto=format&fit=max&s=c27c91ef8517b166794658a094a48642
[1c]:https://www.theguardian.com/artanddesign/gallery/2019/jun/29/the-20-photographs-of-the-week

[2b]:https://i.guim.co.uk/img/media/ad04740a0e3933b5e51e26255035ab3d93dd20a3/0_0_2025_3076/master/2025.jpg?width=1300&quality=45&auto=format&fit=max&dpr=2&s=7bc218e0f697b8e6d193bbc652080292
[2c]:https://www.theguardian.com/artanddesign/2019/jun/30/the-big-picture-tom-wood-liverpool-mothers-go-to-market

[3a]:https://i.guim.co.uk/img/media/7d1688593c21d94ace7a0c178efddf7718f51d9d/0_129_3796_2277/master/3796.jpg?width=3750&quality=85&auto=format&fit=max&s=009e135634cb1f8fdb494ba0ae43c085
[3b]:https://i.guim.co.uk/img/media/30851cf55ebccea4c91dfc8aea7afbd7ec0d2493/0_0_4589_2736/master/4589.jpg?width=1010&quality=45&auto=format&fit=max&dpr=2&s=15580e0160ca24f89d5b0d204f341a87
[3c]:https://www.theguardian.com/news/gallery/2019/jun/30/cappadocia-balloons-and-a-handshake-with-a-dictator-sundays-best-photographs

aw-west commented:
#### Problem:  Saturdays
After further research, I've realised that the Guardian doesn't release a potd on Saturdays [ref][1].  Hence my initial problem.  This means that there might not be a cover photo to the article of `potd_link`.  If there is the original tag `div.i...` will work, if not the replacement tag `div.u...` will find any image on the page.

#### Potential Solution:  Change webpage
Instead of accessing the main guardian page, it might be more reliable to go directly to their [rss feed][2].  Replace lines 171-183 with (with plus asserts)
```
r = requests.get('https://www.theguardian.com/news/series/ten-best-photographs-of-the-day/rss')
soup = BeautifulSoup(r.content, 'lxml')
potd_link = soup.find('guid').contents[0]
```
Or this less explanative soup-free version,
```
r = requests.get('https://www.theguardian.com/news/series/ten-best-photographs-of-the-day/rss')
url = r.text.split('guid')[1].split('>')[1].split('<')[0]
```
On Saturday, the last available `potd_link` is used (Friday).  This doesn't explain the problem I had comparing this morning with this afternoon, but does solve it.

#### Future
`patch-2` is updated for this.  But I guess we should wait a week to test its consistency.


[1]:https://www.theguardian.com/news/series/ten-best-photographs-of-the-day/
[2]:https://www.theguardian.com/news/series/ten-best-photographs-of-the-day/rss
[3]:https://www.theguardian.com/news/gallery/2019/jun/30/cappadocia-balloons-and-a-handshake-with-a-dictator-sundays-best-photographs
